### PR TITLE
MINOR: KTable.count() to only take a selector for key

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
@miguno @ymatsuda 

Hence it does not need to return a key-value, but just the key itself. As for implementation it is mapped to a mapper with (K, V) -> (K1, V).

Also added the equals / hashCode for testing as @miguno suggested.

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java [hunk 1]
```diff
@@ -42,4 +42,21 @@ public KeyValue(K key, V value) {
     public String toString() {
         return "KeyValue(" + key + ", " + value + ")";
     }
+
+    @Override
+    public boolean equals(Object other) {
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java:49]: As an optimization I'd also add as the first check:

``` java
if (other == this) {
  return true;
}
```

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java [hunk 2]
```diff
@@ -42,4 +42,21 @@ public KeyValue(K key, V value) {
...
+        if (other instanceof KeyValue) {
+            KeyValue otherKV = (KeyValue) other;
+
+            return key.equals(otherKV.key) && value.equals(otherKV.value);
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java]: The `KeyValue` class allows null values for `key` and `value` (at least I didn't see input validations such as throwing IAE in its constructor when either key or value are null).  So we must guard against nulls / NPEs here.

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java [hunk 3]
```diff
@@ -42,4 +42,21 @@ public KeyValue(K key, V value) {
...
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        long n = ((long) key.hashCode() << 32) | (long) value.hashCode();
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java]: Same here: NPEs are possible.

Assuming that we do allow null values for `key` and `value`, the tradition says to assume `0` for the hashcode of the null field (see Effective Java, Item 9, p. 47).

One way to achieve that is by using [Objects.hashCode(Object o)](http://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hashCode%28java.lang.Object%29), which returns `0` when `o == null`.

So we could update the code to:

``` java
long n = ((long) Objects.hashCode(key) << 32) | (long) Objects.hashCode(value);
```

That said, we should perhaps use the helper function [`Objects.hash()`](http://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hash%28java.lang.Object...%29) directly, which was introduced in 1.7 (I'd vote for this latter implementation due to simplicity and reusing Java's internal functionality for this):

``` java
@Override
public int hashCode() {
    return Objects.hash(key, value);
}
```

**Comment by miguno** [comment]
I leave the review of the `count` changes to @ymatsuda.

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java (hunk 1 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java:49]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java (hunk 2 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java]: Ah right. Ack.

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java (hunk 3 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java]: ack.

**Comment by guozhangwang** [comment]
@miguno Comments addressed.

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java [hunk 4]
```diff
@@ -42,4 +44,27 @@ public KeyValue(K key, V value) {
     public String toString() {
         return "KeyValue(" + key + ", " + value + ")";
     }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == null)
```
Comment (miguno) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java]: Nitpick:  We don't need to explicitly check for `other == null` -- `instanceof` (which is required for the casting logic anyways) will return false if its first argument is null.

**Comment by miguno** [comment]
General question (since this code will go into Kafka project, and I don't know its policy): What about unit tests, e.g. for the equality semantics of `KeyValue`, which includes verifying `equals` and `hashCode`?  Typically, I'd always use unit tests to ensure that e.g. `equals` is symmetric, transitive, reflexive, and consistent (and that it should work if its argument is null).

**Comment by ymatsuda** [comment]
`count` change looks good to me.

**Comment by guozhangwang** [comment]
@miguo Thanks, I will add the unit tests.

File: streams/src/main/java/org/apache/kafka/streams/KeyValue.java (hunk 4 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KeyValue.java]: Ack.

**Comment by gwenshap** [comment]
LGTM.

As a general note, are we sure that mapping keys and aggregating should be part of the same API call?

why not just: stream.map().count()?

Also, why are we calling the mapper a "selector"? I expected a "selector" to do filtering (i.e. be a predicate).

**Comment by guozhangwang** [comment]
@gwenshap The main motivation is to logically separate aggregation for stream (which is always by key) and for table (which you specify the aggregation key), and the name of "selector" comes from "groupBy" syntax that we plan to extract from KTable. I.e. in the future it will be sth. like:

`ktable.groupBy(mapper).aggregate(aggregator)` instead of `ktable.aggByKey(mapper, aggregator)`.

**Comment by gwenshap** [comment]
so for counting, it will be:
kstream.countByKey()
and 
ktable.groupBy(mapper).count()

What's our user-friendly-explanation for why kstreams and ktables behave differently in this case?

**Comment by guozhangwang** [comment]
It will be

`kstream.count()` which always count by message key; and
`ktable.groupBy().count()` which requires to set the group-by key, since otherwise aggregation is meaningless since in table each key will only have one record logically.

**Comment by guozhangwang** [comment]
I forgot to add the unit test file for this PR, the hotfix one is https://github.com/apache/kafka/pull/960.

**Comment by miguno** [comment]
@guozhangwang: But ktable.count() is possible, too?

**Comment by guozhangwang** [comment]
@miguno Right now ktable.count() needs to provide a "selector" which works as groupBy, what I meant in the previous message is that in the future we will extract it out in a single "groupBy" function.

**Comment by miguno** [comment]
I was referring to a "default" selector so that you can run count() on a ktable without a (required) groupBy. Would this be helpful?

**Comment by guozhangwang** [comment]
What "default" selector do you have in mind?

**Pull request closed without merging**